beam-postgres/Database/Beam/Postgres/Full.hs

```diff
@@ -1,22 +1,100 @@
+{-# LANGUAGE UndecidableInstances #-}
 {-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE TupleSections #-}
 
 -- | Module providing (almost) full support for Postgres query and data
 -- manipulation statements. These functions shadow the functions in
 -- "Database.Beam.Query" and provide a strict superset of functionality. They
 -- map 1-to-1 with the underlying Postgres support.
-module Database.Beam.Postgres.Full where
-
-import           Database.Beam hiding (insertValues)
+module Database.Beam.Postgres.Full
+  ( lockingAllTablesFor_, lockingFor_
+  , PgWithLocking
+  , PgLockedTables, locked_, lockAll_, withLocks_
+  , PgSelectLockingStrength(..), PgSelectLockingOptions(..)
+
+  , insert, insertReturning
+
+  , PgInsertReturning(..)
+  , PgInsertOnConflict(..), PgInsertOnConflictTarget(..)
+  , PgConflictAction(..)
+  , onConflictDefault, onConflict, anyConflict, conflictingFields
+  , conflictingFieldsWhere, conflictingConstraint
+  , onConflictDoNothing, onConflictUpdateSet
+  , onConflictUpdateSetWhere, onConflictUpdateInstead
+  , onConflictSetAll
+
+  , PgUpdateReturning(..)
+  , updateReturning
+
+  , PgDeleteReturning(..)
+  , deleteReturning
+  ) where
+
+import           Database.Beam hiding (insert, insertValues)
 import           Database.Beam.Query.Internal
 import           Database.Beam.Backend.SQL
 import           Database.Beam.Schema.Tables
 
 import           Database.Beam.Postgres.Types
 import           Database.Beam.Postgres.Syntax
 
+import           Control.Monad.Free.Church
+
 import           Data.Monoid ((<>))
 import qualified Data.Text as T
 
+-- * @SELECT@
+
+newtype PgLockedTables s = PgLockedTables [ T.Text ]
+instance Monoid (PgLockedTables s) where
+  mempty = PgLockedTables []
+  mappend (PgLockedTables a) (PgLockedTables b) = PgLockedTables (a <> b)
+
+data PgWithLocking s a = PgWithLocking (PgLockedTables s) a
+instance ProjectibleWithPredicate c syntax a => ProjectibleWithPredicate c syntax (PgWithLocking s a) where
+  project' p mutateM (PgWithLocking tbls a) =
+    PgWithLocking tbls <$> project' p mutateM a
+
+-- | Use with `withLocking_` to lock all tables mentioned in the query
```

Comment:
I think this should be `lockingFor_` right?

Reply:
Yes.. great catch!